man/man1/pmlogpaste.1
'\"macro stdmacro
.\"
.\" Copyright (c) 2020 Red Hat.  All Rights Reserved.
.\"
.\" This program is free software; you can redistribute it and/or modify it
.\" under the terms of the GNU General Public License as published by the
.\" Free Software Foundation; either version 2 of the License, or (at your
.\" option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
.\" or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
.\" for more details.
.\"
.\"
.TH PMLOGPASTE 1 "PCP" "Performance Co-Pilot"
.SH NAME
\f3pmlogpaste\f1 \- paste text into a metric in a PCP archive
.SH SYNOPSIS
\f3pmlogpaste\f1
[\f3\-?\f1]
[\f3\-f\f1 \f2file\f1]
[\f3\-h\f1 \f2hostname\f1]
[\f3\-l\f1 \f2label\f1]
[\f3\-m\f1 \f2metric\f1]
[\f3\-o\f1 \f2outfile\f1]
[\f3\-t\f1 \f2timezone\f1]
.SH DESCRIPTION
.B pmlogpaste
takes input text from a file or the command line, and writes it as
a metric value in a new PCP archive.
This metric value is timestamped with the current time, and is stored
as a string type metric.
.PP
The main purpose of this tool is to take captured output and preserve
this in a PCP archive.
This allows, for example, the output of a benchmark run to be stored
along with performance metrics captured during that run, in a single
archive.
Archives can be merged using the
.BR pmlogextract (1)
utility.
.PP
.B pmlogpaste
uses the
.BR LOGIMPORT (3)
library interfaces internally, which support the creation of archives
from external sources of performance data.
.SH OPTIONS
The available command line options are:
.TP 5
\fB\-f\fR \fIfile\fR, \fB\-\-file\fR=\fIfile\fR
File with input text to be archived, or ``-'' for standard input.
.TP
\fB\-h\fR \fIhostname\fR, \fB\-\-hostname\fR=\fIhostname\fR
Source host name for the archive.
The default value is the local host name, or ``localhost''
if that cannot be obtained.
.TP
\fB\-l\fR \fIlabel\fR, \fB\-\-label\fR=\fIlabel\fR
Add context labels to the generated archive.
This option can be presented multiple times to produce multiple
labels in the archive.
The
.I label
should be in the form
.IR name:value .
.TP
\fB\-m\fR \fImetric\fR, \fB\-\-metric\fR=\fImetric\fR
Metric name for the archive.
The default metric name is
.IR paste.value .
.TP
\fB\-o\fR \fIoutfile\fR, \fB\-\-outfile\fR=\fIoutfile\fR
Archive output file.
The default archive name is
.IR paste .
.TP
\fB\-t\fR \fItimezone\fR, \fB\-\-timezone\fR=\fItimezone\fR
Source timezone for the archive archive.
The default value is the local timezone.
.TP
\fB\-?\fR, \fB\-\-help\fR
Display usage message and exit.
.SH PCP ENVIRONMENT
Environment variables with the prefix \fBPCP_\fP are used to parameterize
the file and directory names used by PCP.
On each installation, the
file \fI/etc/pcp.conf\fP contains the local values for these variables.
The \fB$PCP_CONF\fP variable may be used to specify an alternative
configuration file, as described in \fBpcp.conf\fP(5).
.SH SEE ALSO
.BR PCPIntro (1),
.BR pmcd (1),
.BR pmlogextract (1)
and
.BR LOGIMPORT (3).

.\" control lines not needed for scripts/man-spell


CHANGELOG
pcp-6.2.2 (29 May 2024)
    - Work-in-progress, not yet released.

pcp-6.2.1 (10 April 2024)
    - PMDA additions, enhancements and bug fixes:
	pmdauwsgi: (new) instrumentation from uWSGI servers
	pmdalinux: new hugepage metrics from sysfs (fixed size hugepages)
	pmdalinux: new filesys.uuid and filesys.type metrics
	pmdalinux: new TcpExt metrics from /proc/net/netstat
	pmdalinux: new softnet metrics from newer kernel versions
	pmdalinux: new /proc/vmstat khugepaged metrics
	pmdabpf: updates to add eBPF networking metrics for pcp-atop
	pmdaopenmetrics: fix script error diagnostic cascading exception
    - Client tools and utilities:
	pcp2openmetrics: (new) push PCP metrics in OpenMetrics format
	pmcheck: (new) interrogate and control PCP components
	pmcheck: Redis server detection and agent recommendation
	pmcheck: PostgreSQL server detection and agent recommendation
	pmcheck: uwSGI server detection and agent recommendation
	pcp-atopsar: fix TZ mishandling relating to day-crossover
	pcp-dstat: fix an I/O config typo, sum reads and writes
	pcp-htop: sync latest htop-3.3.0 fixes
	pmrep.conf: use bpf instead of bcc pmda for proc net metrics
	pmrep.conf: proc-essential cosmetic improvements
	pmrep.conf: remove vmeff field from sar -B config
	pmrepconf: use pmrep-internal representations for each metric
    - Server-side utilities and log management scripts:
	pmlogger: bizarre timezone fix (github #1936)
	pmlogger: improve handling of a change in metadata semantics
	pmlogdump: add support for multi-archive contexts
	pmproxy: disable Redis protocol proxying by default
	runaspcp: remove hard-coded "pcp" for user and group
	init scripts: drop conditonal use of setpriv(1) or runuser(1)
    - libpcp, libpcp_pmda, libpcp_mmv, libpcp_web and language bindings:
	libpcp: fix logmeta.c SEGV with V3 archives
	libpcp: fix memleak on repeated __pmFixPMNSHashTab calls
	libpcp: add pmgetopt pmflush() call for warnings
	libpcp: fixes for V3 archives and multi-archive contexts
	libpcp: export __pmLogChangeArchive() interface to tools
	libpcp: fix for PMNS with multi-archive contexts
	libpcp: small derived metrics bug fix (github #1921)
    - Build, infrastructure and packaging updates:
	Linux MX distro support (Debian without systemd)
	riscv: configure pmdabpf_arch value correctly
	tar packages: fix perl modules install in postinstall script
	eBPF updates: add submodule for blazesym, sync to latest
	debian build: add postrm for pcp-zeroconf
	build: removed cppcheck from 'make check', too flakey
	container: Fedora 40 container testing added
	constainer: fixed Fedora base image location (quay.io)
	build: add support for riscv64 to pcp rpm spec files
    - Security Enhanced Linux:
	selinux policy: small tweak for pmie signal handling
    - Documentation:
	man pages: document pmlogger_check and pmlogger_daily SaveLogs
	man pages: man-spell and corrections
	docs: updated donations page with new fiscal sponsor details

pcp-6.2.0 (12 February 2024)
    - Client tools and utilities:
	pmlogger: default to creating version 3 PCP archives
	pmie, pmlogger, pmcd, pmproxy: init script hardening to
	    run less as root:root, more as pcp:pcp and more safely
	pcp-reboot-init: new systemd service for $PCP_RC_DIR setup
	pcp2arrow: new, export metrics to parquet format for pandas
	pcp-atop: deterministic process sort when handling NULLs
	pcp-atop: update to v2.10.0 upstream code
	pcp-htop: update to latest upstream fixes
	pcp-geolocate: function with older versions of python
	pcp-meminfo: resolve possible 'broken pipe' diagnostics
	pcp-ss: fix handling of listening state sockets to match ss
	pcp-ss: fix the UDP UNCONN vs LISTEN state handling
    - PMDA additions, enhancements and bug fixes:
	pmdaproc: improvements for process accounting metrics
	pmdalinux: add pgpromote* and pgdemote* memory metrics
	pmdalinux: add metric sysfs.module.zswap.max_pool_percent
	pmdalinux: add new SNMP metric network.ip.outtransmits
	pmdasummary: send state change 0 (no change) back to pmcd
	pmdapostgresql: support the psycopg3 python module as well
	pmdaopenmetrics: kepler support - live metrics and logging
    - Server-side utilities and log management scripts:
	pmlogrewrite, pmlogredact: fix for missing data volumes
	pmcd: cosmetic change to avoid pipe2 name clash
	runaspcp: new helper utility for unprivileged rc scripts
	pmnsmerge: be more defensive in qsort comparison routine
	pmlogsize: fix qsort comparison function handling of equality
    - libpcp, libpcp_pmda, libpcp_mmv, libpcp_web and language bindings:
	libpcp: pmOpenLog refactor for handling previous log files
	libpcp: new __pmCleanMapDir routine to aid log culling
	libpcp & pmcd: READY-NOT-READY hardening
	libpcp: avoid possible thread issues (coverity)
	libpcp: refactor pmDupContext for thread safety
	libpcp_import: add guard in pmiPutMark to prevent SIGSEGV
    - Build, infrastructure and packaging updates:
	build: changes to re-enable builds for FreeBSD 12, 13, 14
	build: add diffutils rpm dep as pmdaproc script uses diff
	build: add main.yml for building and publishing container
	build: resolve compiler warnings in various tools/libraries
	packaging: use tmpfiles.d method to create PCP_RUN_DIR
	build: packaging changes for openSUSE 15.5 and 15.6
	debian: do not ship empty directory /usr/lib/pkgconfig
	build: el10 systemd-presets for pmcd/pmie/pmlogger enabling
	build: drop rpm dep on bpftool as local version used only
	build: change perl YAML::XS::LibYAML module to YAML::XS
    - Security Enhanced Linux:
	selinux: policy improvements for CentOS/RHEL 7 and 8
    - Documentation:
	docs: Fedora 39 found some tbl(1)/troff(1) issues, fix
	docs: add an explicit code-of-conduct for PCP
	docs: pmns(5) note that metric names must only contain ASCII
	docs: update the BPF README to describe bpftool reality

pcp-6.1.1 (17 November 2023)
    - Client tools and utilities:
	pmie: add -o/--format to control output format with archives

src/pmlogpaste/pmlogpaste.c
/*
 * Copyright (c) 2020 Ashwin Nayak.  All Rights Reserved.
 * Copyright (c) 2020 Red Hat.  All Rights Reserved.
 * 
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the
 * Free Software Foundation; either version 2 of the License, or (at your
 * option) any later version.
 * 
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * for more details.
 */

#include "pmapi.h"
#include "libpcp.h"
#include "import.h"
#include <string.h>

static int  myoverrides(int, pmOptions *);

static pmLongOptions longopts[] = {
    PMAPI_OPTIONS_HEADER("Options"),
    { "hostname", 1, 'h', "HOST", "set hostname" },
    { "timezone", 1, 't', "TZ", "set timezone" },
    { "outfile", 1, 'o', "OUT", "set outfile" },
    { "metric", 1, 'm', "NAME", "set metric name" },
    { "file", 1, 'f', "FILE", "set filename" },
	{ "label", 1, 'l', "LABEL", "set label"},
    PMOPT_HELP,
    PMAPI_OPTIONS_END
};

static pmOptions opts = {
    .flags = PM_OPTFLAG_DONE | PM_OPTFLAG_STDOUT_TZ,
    .short_options = "f:h:m:o:t:l:?",
    .long_options = longopts,
    .short_usage = "[options] archive",
    .override = myoverrides,
};

static char *input;
static size_t input_length;
static char hostname_buffer[MAXHOSTNAMELEN];

/*
 * Append the given buffer to a global (accumulating) string.
 * No assumptions are made about buffer termination (i.e. not
 * explictly NULL terminated here - callers ensure this if it
 * is needed, depending on the context).
 */
static void
append_input(const char *buffer, size_t length)
{
    void	*p;

    if ((p = realloc(input, input_length + length)) == NULL) {
	fprintf(stderr, "%s: out of memory on input\n", pmGetProgname());
	exit(EXIT_FAILURE);
    }
    memcpy(p + input_length, buffer, length);
    input_length += length;
    input = p;
}

/*
 * Reads the contents of a file given an input filename (or '-' for stdin),
 * returns a single buffer.
 */
char *
slurp(const char *filename)
{
    char	buffer[BUFSIZ];
    size_t	length;
    FILE	*file;

    if (strcmp(filename, "-") != 0)
	file = fopen(filename, "r");
    else
	file = stdin;

    if (file) {
	while (!feof(file)) {
	    if ((length = fread(buffer, 1, BUFSIZ, file)) > 0)
		append_input(buffer, length);
	}
	if (strcmp(filename, "-") != 0)
	    fclose(file);
	append_input("\0", 1);	/* ensure string termination */
    }
    return input;
}

/*
 * archives the output from a tool
 */
void 
pmlogpaste(const char *filename, const char *metric,
	   const char *hostname, const char *tz,
	   const char *value, char **labels, int nlabels,
	   struct timespec *timestamp)
{
    int		i, sts;

    if ((sts = pmiStart(filename, 0)) < 0) {
	fprintf(stderr, "%s: error starting log import: %s\n",
			pmGetProgname(), pmiErrStr(sts));
	exit(EXIT_FAILURE);
    }

    if ((sts = pmiSetHostname(hostname)) < 0) {
	fprintf(stderr, "%s: error setting log hostname: %s\n",
			pmGetProgname(), pmiErrStr(sts));
	exit(EXIT_FAILURE);
    }

    if ((sts = pmiSetTimezone(tz)) < 0) {
	fprintf(stderr, "%s: error setting log timezone: %s\n",
			pmGetProgname(), pmiErrStr(sts));
	exit(EXIT_FAILURE);
    }

    if ((sts = pmiAddMetric(metric, PM_ID_NULL, PM_TYPE_STRING, PM_INDOM_NULL,
			    PM_SEM_DISCRETE, pmiUnits(0, 0, 0, 0, 0, 0))) < 0) {
	fprintf(stderr, "%s: error adding metric descriptor: %s\n",
			pmGetProgname(), pmiErrStr(sts));
	exit(EXIT_FAILURE);
    }

    if ((sts = pmiPutValue(metric, NULL, value)) < 0) {
	fprintf(stderr, "%s: error adding metric value: %s\n",
			pmGetProgname(), pmiErrStr(sts));
	exit(EXIT_FAILURE);
    }

    for (i = 0; i < nlabels; i++) {
	char 	*temp = strdup(*(labels + i));
	char	*name = strtok(temp, ":");
	char	*data = strtok(NULL, ":");

	if (name == NULL) {
	    fprintf(stderr, "%s: invalid label token %s: %s\n",
			pmGetProgname(), temp, pmiErrStr(sts));
	    exit(EXIT_FAILURE);
	}
	if ((sts = pmiPutLabel(PM_LABEL_CONTEXT, PM_ID_NULL, PM_IN_NULL, name, data)) < 0) {
	    fprintf(stderr, "%s: error adding label %s:%s: %s\n",
			pmGetProgname(), name, data, pmiErrStr(sts));
	    exit(EXIT_FAILURE);
	}
	free(temp);
    }
    
    if ((sts = pmiWrite(timestamp->tv_sec, timestamp->tv_nsec / 1000)) < 0) {
	fprintf(stderr, "%s: error writing archive: %s\n",
			pmGetProgname(), pmiErrStr(sts));
	exit(EXIT_FAILURE);
    }

    if ((sts = pmiEnd()) < 0) {
	fprintf(stderr, "%s: error in ending log writing: %s\n",
			pmGetProgname(), pmiErrStr(sts));
	exit(EXIT_FAILURE);
    }
}

/*
 * pmlogpaste has a few options which do not follow the defacto standards
 */
static int
myoverrides(int opt, pmOptions *options)
{
    if (opt == 'h' || opt == 't')
	return 1;	/* we've claimed these, inform pmGetOptions */
    return 0;
}

int 
main(int argc, char *argv[])
{
    int		opt, exitsts;
    char	*filename = NULL;
    char	*metric = NULL;
    char	*outfile = NULL;
    char	*hostname = NULL;
    char	*tz = NULL;
    char	**labels = NULL;
    int 	nlabels = 0;
    struct timespec timestamp = {0};

    while ((opt = pmGetOptions(argc, argv, &opts)) != EOF) {
	switch(opt) {

	case 'f':
	    filename = opts.optarg;
	    break;

	case 'm':
	    metric = opts.optarg;
	    break;

	case 'o':
	    outfile = opts.optarg;
	    break;

	case 'h':
	    hostname = opts.optarg;
	    break;

	case 't':
	    tz = opts.optarg;
	    break;
	
	case 'l':
	    nlabels++;
	    labels = (char **)realloc(labels, nlabels * (sizeof(char *)));
	    if (labels == NULL)
		pmNoMem("pmlogpaste.labels", nlabels * sizeof(char *), PM_FATAL_ERR);
	    *(labels + nlabels - 1) = opts.optarg;
	    break;

	case '?':
	    break;

	default:
	    opts.errors++;
	    break;
	}
    }

    if (filename && opts.optind < argc) {
	fprintf(stderr,
		"%s: -f/--file cannot be used with command line input\n",
		pmGetProgname());
	opts.errors++;
    }

    if (opts.errors || (opts.flags & PM_OPTFLAG_EXIT)) {
	exitsts = !(opts.flags & PM_OPTFLAG_EXIT);
	pmUsageMessage(&opts);
	exit(exitsts);
    }

    if (filename == NULL && opts.optind < argc) {
	for (opt = opts.optind; opt < argc; opt++) {
	    append_input(argv[opt], strlen(argv[opt]));
	    append_input(" ", 1);
	}
	input[input_length-1] = '\0';
    } else {
	if (filename == NULL)
	    filename = "-";	/* read from standard input when nowhere else */
	slurp(filename);
    }

    if (metric == NULL)
	metric = "paste.value";		/* default metric name */

    if (outfile == NULL)
	outfile = "paste";		/* default archive name */

    if (tz == NULL)
	tz = __pmTimezone();

    if (__pmGetTimespec(&timestamp) < 0)	/* high resolution timestamp */
	timestamp.tv_sec = time(NULL);

    if (hostname == NULL) {
	if ((gethostname(hostname_buffer, sizeof(hostname_buffer))) < 0)
	    hostname = "localhost";
	else
	    hostname = &hostname_buffer[0];
    }

    pmlogpaste(outfile, metric, hostname, tz, input, labels, nlabels, &timestamp);

    if (labels)
	free(labels);
    return 0;
}
